feat!: Improve PM caching and remote referencing

nderjung opened the pull request:
### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This large refactor makes improvements in KraftKit's handling of local OCI image caching towards improving performance and reliability.

nderjung committed:
fix(oci): Prevent pointer exceptions for nil-values of `query`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Set KraftKit's version at point-of-`Save`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(oci): Implement OCI Package Manager `Update` method

This reduces code duplication by abstracting the core functionality of
the update procedure into an internal `update` method which can be
re-used by both the publically exposed `Update` interface method as well
as the `Catalog` interface method which can request to do an "update"
on-the-fly when the query option for an update is set to true.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci)!: Deprecate `DigestExists` in favor of `DigestInfo`

This commit reworks `DigestExists` into a new method, `DigestInfo`,
which returns rich output regarding the creation time, update time, and
size of the digest.  Naturally, returning a non-nil value of this
information indicates that the digest exists and so all previous
references to `DigestExists` are updated appropriately.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci)!: Move `PlatformChecksum` into utils package

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Correctly hydrate manifest from descriptor's digest

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
style(oci): Homogenize log fields by using `WithField` over `WithFields`

This commit transforms entries in the `log` package from using:

```go
log.G(ctx).WithFields(...)
```

To:

```go
log.G(ctx).WithField(...)
```

Simply ensuring consistency between all entries.

Furthermore, additional log entries are injected throughout to aid in
debug/tracability.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Introduce interface method `PulledAt` to `pack.Package`

This well-known metadata attribute can be used to determine whether the
package has been previously pulled along with when.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Remove old "commented-out" code

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Introduce `Unpack` interface method to `pack.Package`

This commit introduces a new method to all package implementations,
`Unpack`, which allows for the separation of pulling and unpackaging
into two distinct steps.  Note that for now, this is only consumed at
the application-level with the `runner_package.go` (`kraft run`) and
therefore OCI packages and not fully implemented or Manifest-type
packages.

Providing such a distinction speeds up the instantiation of pre-pulled
packages representing applications.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(handler): Overwrite entries representing the same plat checksum

This commit fixes an issue where a reference to a manifest may exist
locally and after a pull of the same manifest, said same entry needs to
be removed and updated with all associated layers (thus also changing
the digest of the manifest).

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Pass context to remote index options

This fixes an issue where context may be cancelled and the operation of
looking up the remote index does not stop.  Passing the current context
via `WithContext` remedies this issue.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Improve peformance of `NewIndexFromRef`

Use `ResolveIndex` over `ListIndexes` to prevent unnecessary iteration.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor!: Distinguish between "local" and "remote" queries

This commit deprecates the package manager's query option for
`WithUpdate` and splits it into two distinct options: `WithLocal` and
`WithRemote` where local queries are by default `true` and `WithLocal`
can be used to set local queries to false and prevent local lookups.

Accordingly, the `Catalog` method for OCI and Manifest package managers
are updated to reflect this.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Remove unused `WithPlatform`

Upon further inspection, this `remote.Option` is never utilized by the
underlying package and so its inclusion is not necessary.  Remove.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(cache): Introduce `RemoteIndex`

`RemoteIndex` is a wrapper for `v1.RemoteIndex` which caches requests
in-memory for previously requested indexes.  This aids in reducing the
number of lookup calls to the same index.  Since `v1.WithPlatform` is not
respected, a valid lookup will have had any additional options, such as
`WithTransport` and `WithAuth`, fully satisfied.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(cache): Introduce `RemoteImage`

`RemoteImage` is a wrapper for `v1.Image` which caches requests
in-memory for previously requested image.  This aids in reducing
the number of lookup calls to the same images.  Since
`v1.WithPlatform` is not respected, a valid lookup will have had
any additional options, such as `WithTransport` and `WithAuth`,
fully satisfied.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci): Improve `NewPackageFromOCIManifestDigest`

This commit reworks parts of `NewPackageFromOCIManifestDigest` to use
parallelizm when looking up indexes containing multiple manifest images
which can be handled concurrently as well correctly seeding unknown
manifests that could not be determined locally.  This generally improves
the performance and reliability of this method when passing in
references.

Additionally, the recently introduced caching of v1(Index)Image is used
accordingly.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(oci): Handle partial (non-distributable) OCI layers

This commit reworks internal mechanisms in the OCI package such that it
is capable of handling packages which simply _reference_ a layer without
having to pull or upload it.  This is both an optimization step in terms
of preventing users who are simply re-packaging unikernels with new
rootfilesystem layers to download and re-upload the same kernel layer as
well as redundant cache coping mechanism such that remote references to
kernel layers are kept up-to-date and re-packaging ensures the latest
reference.

Numerous changes were necessary across the OCI package manager to make
this possible, including:

- keeping an in-memory copy of previously instantiated `ociPackage`
  structure;
- preventing error-outs from not found layers (which are now logged as
  non-distributable);
- better handling of saving and updating of the `manifest` and `index`
  internal structures if and when changes to the manifest and index
  exactly occur; and,
- "re-tagging" non-local, "non-distributable" OCI layers (in compliance
  with the OCI Distribution Specification) during the `Push` procedure.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Fix possible nil-pointer exception

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci)!: Toggle whether to delete index dependencies

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(handler): Improve detection of local cache in `PullDigest`

This commit reworks parts of the `DirectoryHandler`'s `PullDigest`
interface method to first perform a check for the digest and if it
exists locally.  Additionally, use of the v1 remote index/image cacher
noticibly reduces the number of remote lookups.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(handler): Include known local manifest if digest is identical

Bypass the platform checksum check if the manifest digest is the
same as the requested manifest, this can be safely stored in the
index.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(handler): Always remove the manifest

This commit fixes an issue where the manifest is not deleted if any of
its dependencies have failed to be removed.

Note: this may lead to lingering digests, but during testing I could not
detect this.  Instead, errors were generally to do with missing
dependencies as opposed to inability to remove them.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(oci): Expose OCI package size in list of columns

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Introduce `Save` interface method to `pack.Package`

This allows the package's current state to be saved.  For now, OCI
package manager is implemented.  Ultimately this allows for ephemeral
packages that do not need to be saved to disk to be utilized in more
high-level code usages.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Provide default tag if not provided

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(handler): Fix containerd's PullDigest based on `WithPlatform`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Utilize caching via `RemoteIndex`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Include missing copyright header notice

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Remove redundant brackets

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Correct and improve return error messages

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(pkg): Rework the runtime packager to support non-pulls

This commit reworks the subcommand for `kraft pkg` when a `Kraftfile`
specifies a `runtime` element with a new flag `--no-pull`.  Ultimately
this stems from the rework done in the package manager and specifically
the OCI package manager and its ability to reference non-distributable
OCI image layers.

To accommodate this change, the packager is updated to use the `Save`,
`IsPulled` and `Remote`/`Local` interface methods from the
`pack.Package` and performs an appropriate search before re-packaging
the filesystem with a remote package reference.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(deploy): Set `NoPull` to true for kraft cloud images

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(tui): Do not print the `textLeft` if in norender and no hide mode

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(pull): Better handling of `kraft pkg pull`

This commit partially reworks `kraft pkg pull` to factor out the
embedded package manager which was previously part of `pmQuery`.
Simultaneously, it migrates all pullable packages into a final, separate
process tree, rather than in-lining the pull after the find.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(update): Fix incorrect package due to Go closures

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Support updating all package managers

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci): Deprecate github.com/genuinetools/reg

This package has not been updated in 4 years and is starting to cause
issues with outdated packages causing conflicts.  Instead, use
github.com/google/go-containerregistry which has similar methods.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(run): Prevent pull step if package exists

Thanks to the introduction of `PulledAt` interface method of
`pack.Package`, we can now omit the pull step entirely if it already
exists locally.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(utils): Show package pull time

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Speed up exact index `Catalog` search

If a fully qualified OCI reference is possible for a local look up, use
the `ResolveIndex`, a 1-op oepration, as opposed to iterating through
each index via `ListIndexes`.  This speeds up the return values from the
`Catalog` method for the OCI package manager.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on oci/handler/directory.go:
```diff
@@ -778,9 +778,15 @@ func (handle *DirectoryHandler) DeleteManifest(ctx context.Context, fullref stri
+			WithField("digest", dgst.String()).
+			Trace("deleting manifest")
+
+		os.RemoveAll(manifestPath)
```
Print error if returned, would maybe help in permission error cases

craciunoiuc commented on internal/cli/kraft/pkg/packager_kraftfile_runtime.go:
```diff
@@ -110,95 +113,181 @@ func (p *packagerKraftfileRuntime) Pack(ctx context.Context, opts *PkgOptions, a
+	// Remove the cached runtime package reference if it was not previously
+	// pulled.
+	if !pulled {
+		defer runtime.Delete(ctx) //nolint:errcheck
```
don't hide it :D

You can do a `defer func() {` And also log the error if it pops up

nderjung replied:
This is specifically hidden because it is a fail-safe, catch-all.  I can however add it to the trace logs to keep it available, but it will almost always fire.

nderjung committed:
fix(pull): Fix check to only pull template components

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(pull)!: Rework `--force-cache` to become `--update`

By default, `kraft pkg pull` should use local information to determine
what to pull.  When relevant package components cannot be found, a
remote search is necessary.  Provide this functionality by deprecating
the `--force-cache` flag in favour of the `--update` flag which signals
to the package manager's `Catalog` method to perform said remote seach.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(pull): Do not log a warning, explicitly error on not-found packages

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(manifest): Correct when to save the origin of the manifest

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(manifest): Downgrade warn-level message to trace-level

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(manifest): Expose origin in the list of columns

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Ensure to check each individual package from a remote update

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(pkg)!: Rename `Manager` to `Format` and use `-f` short flag

Better name reflection of the actual attribute.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(processtree): Introduce new option `WithHideError`

This new option allows toggling the UI to hide any logs and top-level
error messaging.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(pull): Reliably pull with local-first, remote-second attempt

This re-attempts the `Catalog` search by first performing a local query
for references.  Since `kraft pkg pull` is used to fetch a remote
package, re-attempting by connecting to remote sources makes sense to
the user.  Adjust the workflow to accommodate for this second attempt.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(handler): Parallelize pulling manifests

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(manifest): Fix possible nil-pointer exception

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on oci/manager.go:
```diff
@@ -492,19 +492,45 @@ func (manager *ociManager) Catalog(ctx context.Context, qopts ...packmanager.Que
+							log.G(ctx).
+								WithField("want", fmt.Sprintf("%s:%s", qname, qversion)).
+								WithField("got", fullref).
+								Trace("skipping manifest: name does not match")
```
`Trace("skipping manifest: name and version do not match")`

to differentiate from the one below which has the same message

nderjung committed:
fix(pull): Remove unused `--with-deps` and `--no-deps` flags

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Better handling of partial indexes

This commit migrates the functionality of
`NewPackageFromOCIManifestDigest` which handles remote access of an
index and its manifest into `newIndexAndManifestFromRemoteDigest`.  This
method is then used during a re-attempt in accessing an index and its
manifest if a local reference cannot be instantiated.  This increases
the overal resilience of the exposed method.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(pull): Add Kraftfile runtime to list of pull queries

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(pull): Add early error message for no queries

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc reviewed (approved):
Mercy be upon us.

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
